src/peer_select.cc

```diff
@@ -1083,6 +1083,8 @@ PeerSelector::addSelection(CachePeer *peer, const hier_code code)
         // There can be at most one PINNED destination.
         // Non-PINNED destinations are uniquely identified by their CachePeer
         // (even though a DIRECT destination might match a cache_peer address).
+        // XXX: We may still add duplicates because the same peer could have
+        // been removed from `servers` already (and given to the requestor).
         const bool duplicate = (server->code == PINNED) ?
                                (code == PINNED) : (server->_peer == peer);
```

Comment:
FTR: This newly discovered XXX may result in occasional violation of the documented peer selection [algorithm](https://wiki.squid-cache.org/Features/LoadBalance#Overall_peer_selection_logic), but its effects, if any, should be very rare and mild in most environments.

Reply:
FYI: That description of the problem as being minor for me indicates that this is a TODO, rather than XXX importance.

Reply:
I speculate that it is minor in most environments, but it is a bug that may badly screw up some deployments. Said that, if you insist on changing this to TODO, I am not going to fight about it.

Reply:
Okay. Then please make the comment clarify why having a peer be tried twice composes being "a problem, often a serious one" (per our current ad-hoc policy).

I do not recall the previous behaviour of peer selection when there wold often be whole series of duplicate entries (dozens at times) being a big problem. All it did was lead to some connections being slower, and some user complaints about non-contactable sites when the origin was provably broken. IIRC, the de-dup logic was just a performance optimization not a bug fix.

Reply:
> Then please make the comment clarify why having a peer be tried twice composes being "a problem, often a serious one"

You are asking me to clarify why missing a duplicate in the code that claims to "Bail on a duplicate" and computes a variable called "duplicate" is "a problem, often a serious one". Sigh.

Can we just delete this XXX instead? I do not want to waste even more time on this. If this XXX must stay to pass your review, then please suggest specific wording that would make you happy.

Reply:
Done at 31c8392 to facilitate the PR progress.